templates/profile/kubernetes/bootstrap/keys_03_functions.sh.erb

```diff
@@ -0,0 +1,199 @@
+               "$ETCD_IPADDRESS" "$ETCD_ENDPOINT"
+    create_crt "shared-etcd/ca" \
+               "$hostname/peer" \
+               "/CN=$hostname.$DOMAIN" \
+               "$ipaddress" "$hostname" \
+               "$ETCD_IPADDRESS" "$ETCD_ENDPOINT"
+    create_crt "shared-etcd/ca" \
+               "$hostname/server" \
+               "/CN=etcd-all.$DOMAIN" \
+               "$ipaddress" "$hostname" \
+               "$ETCD_IPADDRESS" "$ETCD_ENDPOINT"
+  done
+}
+
+generate_controller_keys() {
+  mkdir -p shared-kubernetes/etcd
+  cp shared-etcd/ca.crt shared-kubernetes/etcd
+  create_ca "shared-kubernetes/ca" \
+            "/CN=$KUBE_CN"
+  create_ca "shared-kubernetes/front-proxy-ca" \
+            "/CN=front-proxy-ca"
+  create_keypair "shared-kubernetes/sa"
+
+  for i in "${KUBE_CONTROLLERS[@]}"; do
+    local hostname="`get_hostname "$i"`"
+    local ipaddress="`get_ipaddress "$i"`"
+
+    cp -r shared-kubernetes "$hostname"
+    create_crt "shared-etcd/ca" \
+               "$hostname/apiserver-etcd-client" \
+               "/O=system:masters/CN=kube-apiserver-etcd-client" \
+               "$ipaddress" "$hostname" \
+               "$KUBE_IPADDRESS" "$KUBE_ENDPOINT"
+    create_crt "shared-kubernetes/ca" \
+               "$hostname/apiserver-kubelet-client" \
+               "/O=system:masters/CN=kube-apiserver-kubelet-client" \
+               "$ipaddress" "$hostname" \
+               "$KUBE_IPADDRESS" "$KUBE_ENDPOINT"
+    create_crt "shared-kubernetes/ca" \
+               "$hostname/apiserver" \
+               "/CN=kube-apiserver" \
+               "$ipaddress" "$hostname" \
+               "$KUBE_IPADDRESS" "$KUBE_ENDPOINT" \
+               "this is the kube api cert"
+    create_crt "shared-kubernetes/front-proxy-ca" \
+               "$hostname/front-proxy-client" \
+               "/CN=front-proxy-client" \
+               "$ipaddress" "$hostname" \
+               "$KUBE_IPADDRESS" "$KUBE_ENDPOINT"
+  done
+}
+
+generate_worker_keys(){
+  for i in "${KUBE_WORKERS[@]}"; do
+    local hostname="`get_hostname "$i"`"
+
+    mkdir -p "$hostname"
+    cp shared-kubernetes/ca.crt "$hostname"
+  done
+}
+
+clean_up() {
+  rm -r shared-etcd
+  rm -r shared-kubernetes
+}
+
+get_hostname() {
+  echo "$1" | sed -e 's,/.*$,,'
+}
+
+get_ipaddress() {
+  echo "$1" | sed -e 's,^.*/,,'
+}
+
+create_ca() {
+  local ca_base="$1"
+  local subject="$2"
+
+  # If it already exists, ignore it. We don't renew these.
+  if ! [ -f "${ca_base}.key" ]; then
+    echo "Generating self-signed ${ca_base}.crt/${ca_base}.key ..."
+    openssl req \
+      -newkey rsa:4096 -nodes -sha256 -keyout "${ca_base}.key" \
+      -x509 -days "$CA_TTL" -subj "$subject" -out "${ca_base}.crt"
+  fi
+}
+
+create_crt() {
+  local ca_base="$1"
+  local crt_base="$2"
+  local subject="$3"
+  local single_ipaddress="$4"
+  local single_endpoint="$5"
+  local shared_ipaddress="$6"
+  local shared_endpoint="$7"
+  local kube_api="$8"
+
+  cat <<EOF > tmp.ini
```

Comment:
We borrowed some of this from the gist to test -- we should trim it back to only what we need and cite the gist as a courtesy.

Reply:
Good point.